plugin/db/mysql/rollback.go

```diff
@@ -0,0 +1,140 @@
+package mysql
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// GetRollbackSQL generates the rollback SQL for the list of binlog events in the reversed order.
+func (txn BinlogTransaction) GetRollbackSQL(columnNames []string) (string, error) {
+	if len(txn) == 0 {
+		return "", nil
+	}
+	var sqlList []string
+	// Generate rollback SQL for each statement of the transaction in reversed order.
+	// Each statement may have multiple affected rows in a single binlog event. The order between them is irrelevant.
+	for i := len(txn) - 1; i >= 0; i-- {
+		e := txn[i]
+		if e.Type != WriteRowsEventType && e.Type != DeleteRowsEventType && e.Type != UpdateRowsEventType {
+			continue
+		}
+		sql, err := e.getRollbackSQL(columnNames)
+		if err != nil {
+			return "", err
+		}
+		sqlList = append(sqlList, sql)
+	}
+	return strings.Join(sqlList, "\n"), nil
+}
+
+var (
+	reDMLBlockPrefix = regexp.MustCompile("(?m)^### ")
+	reInsertInto     = regexp.MustCompile("(?m)^INSERT INTO")
+	reDeleteFrom     = regexp.MustCompile("(?m)^DELETE FROM")
+)
+
+func (e *BinlogEvent) getRollbackSQL(columnNames []string) (string, error) {
+	// 1. Remove the "### " prefix of each line.
+	body := reDMLBlockPrefix.ReplaceAllLiteralString(e.Body, "")
+
+	// 2. Switch "DELETE FROM" and "INSERT INTO".
+	// 3. Replace "WHERE" and "SET" with each other.
+	// 4. Replace "@i" with the column names.
+	// 5. Add a ";" at the end of each row.
+	var err error
+	switch e.Type {
+	case WriteRowsEventType:
+		body = reInsertInto.ReplaceAllLiteralString(body, "DELETE FROM")
```

Comment:
We should only replace the first instead of all as values could contain such strings.

Reply:
If an original SQL inserts multiple rows, there will be multiple "INSERT INTO" in the binlog event.